### PR TITLE
[RELAY/PASS] Change the return value convention of gradient

### PR DESCRIPTION
**Pull request by ZihengJiang**
Change the return value convention from `(res, (grad0, ...))` to `(res, grad0, ...)` to void nested tuple.

@MarisaKirisame @tqchen @jroesch @reminisce

### CODE REVIEW

**Comment by MarisaKirisame** [comment]
I am a bit against this change. while it seems to improve usability, it harm composability.
To be more specific, we could do x.1 to take all the gradient, and put it into a tuple, so we can build a gradient descent/adam function in relay.
After this change, it is impossible, so how can we move the training loop into relay?
I do agree that improving usability is important, so maybe a better way is to keep the old version (as gradient_raw) alongside the new version.

**Comment by tqchen** [comment]
Given the current convention, it is still easy to extract the gradient values via outputs[1:]. The main trade-off here is the generality v.s. the convention to favor non-nesting tuples. Non-nesting tuples are easier to optimize in general and work well for old graph runtime, and so I would recommend to keep them when possible.

**Comment by jroesch** [comment]
I agree with Marisa. We have discussed adding a flattening pass for tuples, but I feel we should do that optimization separately from the interface. To me the interface is that we return the forward evaluation AND the gradients as two separate entities.

**Comment by MarisaKirisame** [comment]
to be more clear, i think we can have both by having two interface - one general one, for whoever need to write generic code (such as a generic gradient descend), and one where where it is flatten and easier to use.
we cant go from the flatten one to the unflatten one purely inside relay world(we have to goto python)

**Comment by ZihengJiang** [comment]
I feel that it is just about convention and keep nested tuple seems no direct benefit?

**Comment by ZihengJiang** [comment]
> we cant go from the flatten one to the unflatten one purely inside relay world(we have to goto python)

@MarisaKirisame do you mean we cannot express `ret[1:]` in relay?

**Comment by MarisaKirisame** [comment]
@ZihengJiang when ret is a tuple, exactly.

**Comment by tqchen** [comment]
OK, given the current discussion and suggestions from @MarisaKirisame and @jroesch , let us still keep the original signature, but provide a pass tuple_flatten that flattens the nested tuple into the flattened one

**Comment by MarisaKirisame** [comment]
@ZihengJiang can you fix foward to forward? i can do it too.
